Replace TinyMCE with QuillJS 2.0.3

fballiano opened the pull request:
## Summary

This PR completely removes TinyMCE from Maho and replaces it with QuillJS 2.0.3.

### Why remove TinyMCE?
- The GPL-licensed version of TinyMCE is incompatible with Maho's OSL-3.0 license
- The MIT-licensed version (6.8) is unmaintained and has unpatched security vulnerabilities and mark a couple of our repositories as "problematic", this is a no go

## Changes

### Core Implementation
- **Removed TinyMCE**: Deleted all TinyMCE files (`/public/js/tinymce/` and `/public/js/mage/adminhtml/wysiwyg/tinymce/`)
- **Added QuillJS 2.0.3**: Loaded from CDN (cdn.jsdelivr.net) in admin head template
- **New setup script**: Created `/public/js/mage/adminhtml/wysiwyg/quill/setup.js` with full QuillJS integration
- **Custom styling**: Added `/public/js/mage/adminhtml/wysiwyg/quill/quill-maho.css` to match Maho admin theme

### Feature Parity
- ✅ Widget insertion support with custom toolbar button
- ✅ Variable insertion support with custom toolbar button  
- ✅ Media browser integration for images/files
- ✅ Show/Hide editor toggle functionality

### Configuration Updates
- Removed TinyMCE composer check from `Mage_Cms_Model_Wysiwyg_Config`
- Removed skin configuration (`cms/wysiwyg/skin`) and related model
- Updated layout XML to load QuillJS assets with `can_load_wysiwyg` condition
- Updated all PHP blocks to set both `setCanLoadTinyMce()` and `setCanLoadWysiwyg()` for compatibility

## Breaking Changes

1. **TinyMCE is no longer available** - Any custom TinyMCE plugins or configurations will need to be rewritten for QuillJS
2. **Removed configuration option** - `cms/wysiwyg/skin` no longer exists
3. Encoding/decoding directives have been removed for simplicity (it was incomplete before already)
4. **JavaScript API changes** - Custom code using TinyMCE API must be updated to use QuillJS API

## Testing

Test WYSIWYG functionality in all admin areas:
- [ ] CMS Pages (content editor)
- [ ] CMS Static Blocks  
- [ ] Product descriptions (short and full)
- [ ] Category descriptions
- [ ] Newsletter templates
- [ ] Email templates
- [ ] Widget insertion
- [ ] Variable insertion
- [ ] Media/image browser
- [ ] Show/Hide editor toggle

justinbeaty commented:
Oh nice, I will review this asap!

justinbeaty commented:
What do you think about this patch? We could then remove all of the `$this->getLayout()->getBlock('head')->setCanLoadTinyMce(true);` lines.

```diff
diff --git a/app/code/core/Mage/Page/Block/Html/Head.php b/app/code/core/Mage/Page/Block/Html/Head.php
index 4ebd1a1be4..e2179ceeb8 100644
--- a/app/code/core/Mage/Page/Block/Html/Head.php
+++ b/app/code/core/Mage/Page/Block/Html/Head.php
@@ -14,7 +14,6 @@
  * @method $this setCanLoadCalendarJs(bool $value)
  * @method $this setDescription(string $value)
  * @method $this setKeywords(string $value)
- * @method $this setCanLoadTinyMce(bool $value)
  * @method $this setCanLoadWysiwyg(bool $value)
  */
 class Mage_Page_Block_Html_Head extends Mage_Core_Block_Template
@@ -561,4 +560,14 @@ class Mage_Page_Block_Html_Head extends Mage_Core_Block_Template
             $this->_data['items'] = $newItems;
         }
     }
+
+    public function setCanLoadTinyMce(bool $value): self
+    {
+        return $this->setData('can_load_wysiwyg', $value);
+    }
+
+    public function getCanLoadTinyMce(): bool
+    {
+        return (bool) $this->getData('can_load_wysiwyg');
+    }
 }
```

justinbeaty commented:
I'm actually having trouble inserting images, widgets, etc. I hard refreshed, etc.

About the directive encoding, are you sure we can get rid of it? When I edited a page with an image, the markup became this:
```html
<img src="{{media url=" wysiwyg="" 404-banner-3.webp"}}"="" alt="404bannerimage">
```

That got pretty broken because of the nested double quotes.

fballiano commented:
Images dont work yet but widgets and variables should (im out now sorry)

Il giorno sab 24 mag 2025 alle 22:34 Justin Beaty ***@***.***>
ha scritto:

> *justinbeaty* left a comment (MahoCommerce/maho#174)
> <https://github.com/MahoCommerce/maho/pull/174#issuecomment-2907084708>
>
> I'm actually having trouble inserting images, widgets, etc. I hard
> refreshed, etc.
>
> About the directive encoding, are you sure we can get rid of it? When I
> edited a page with an image, the markup became this:
>
> <img src="{{media url=" wysiwyg="" 404-banner-3.webp"}}"="" alt="404bannerimage">
>
> That got pretty broken because of the nested double quotes.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MahoCommerce/maho/pull/174#issuecomment-2907084708>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAG6DL7S4X5EZS2TYHG7MKD3ADQWPAVCNFSM6AAAAAB5Z2FSK6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMBXGA4DINZQHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

justinbeaty commented:
You're right, the widgets and variables are working, sorry not sure how I missed it.

Regarding the encoding, it definitely is a problem. For example, I created a new page, switched to plaintext mode, and entered this:

```html
<a href="mailto:{{config path="trans_email/ident_custom1/email"}}">Email Me</a>
```

Then saved, and switched to the plaintext again and it turned into:

```html
<p><a href="mailto:{{config path=" trans_email="" ident_custom1="" email"}}"="">Email Me</a></p>
```

fballiano commented:
> What do you think about this patch? We could then remove all of the `$this->getLayout()->getBlock('head')->setCanLoadTinyMce(true);` lines.
> 
> ```diff
> diff --git a/app/code/core/Mage/Page/Block/Html/Head.php b/app/code/core/Mage/Page/Block/Html/Head.php
> index 4ebd1a1be4..e2179ceeb8 100644
> --- a/app/code/core/Mage/Page/Block/Html/Head.php
> +++ b/app/code/core/Mage/Page/Block/Html/Head.php
> @@ -14,7 +14,6 @@
>   * @method $this setCanLoadCalendarJs(bool $value)
>   * @method $this setDescription(string $value)
>   * @method $this setKeywords(string $value)
> - * @method $this setCanLoadTinyMce(bool $value)
>   * @method $this setCanLoadWysiwyg(bool $value)
>   */
>  class Mage_Page_Block_Html_Head extends Mage_Core_Block_Template
> @@ -561,4 +560,14 @@ class Mage_Page_Block_Html_Head extends Mage_Core_Block_Template
>              $this->_data['items'] = $newItems;
>          }
>      }
> +
> +    public function setCanLoadTinyMce(bool $value): self
> +    {
> +        return $this->setData('can_load_wysiwyg', $value);
> +    }
> +
> +    public function getCanLoadTinyMce(): bool
> +    {
> +        return (bool) $this->getData('can_load_wysiwyg');
> +    }
>  }
> ```

what do you think about https://github.com/MahoCommerce/maho/pull/174/commits/8887591d9611018b331a0c3be5236a4693ba1db6?

fballiano commented:
will continue to work on this tomorrow :-)

fballiano commented:
> Regarding the encoding, it definitely is a problem. For example, I created a new page, switched to plaintext mode, and entered this:
> 
> ```
> <a href="mailto:{{config path="trans_email/ident_custom1/email"}}">Email Me</a>
> ```
> 
> Then saved, and switched to the plaintext again and it turned into:
> 
> ```
> <p><a href="mailto:{{config path=" trans_email="" ident_custom1="" email"}}"="">Email Me</a></p>
> ```

riiiiight I thought it was only about the visuals but it's not, could you check the last version? it's not beautiful but it should mostly work

fballiano commented:
nah, I have to go back to the original implementation cause it has to be backward compatible :-\

fballiano commented:
I'm moving the development to branch quilljs3 because this one is broken

fballiano commented:
> Hm, yeah. I recall looking at the original implementation in #87 and not touching it because it was hard to understand. I think it was working okay, but I don't use the wysiwyg very much.

:-D and it was too complicated to just reintroduce that commit back so I kinda restarted but it's better cause I think it will be cleaner in the end. I'll close this PR and will open another one for that other branch soon :-)
